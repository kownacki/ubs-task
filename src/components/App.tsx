import React, { FC } from 'react';
import { Provider } from 'react-redux';
import '../index.css';
import { Header } from './Header';
import { Contents } from './Contents';
import { store } from '../redux/store';

export const App: FC = () => {
  return (
    <Provider store={store}>
      <Header />
      <Contents />
    </Provider>
  );
};
