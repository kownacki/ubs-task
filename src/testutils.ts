export const categoriesStub = [
  {
    'id': 0,
    'name': 'Cycling'
  },
  {
    'id': 1,
    'name': 'Hiking'
  },
  {
    'id': 2,
    'name': 'Cooking'
  }
];

export const coordinatorsStub = [
  {
    'id': 0,
    'name': 'Daniel',
    'lastname': 'Mitchell',
    'email': 'daniel.mitchell@hussa.rs'
  },
  {
    'id': 1,
    'name': 'Albert',
    'lastname': 'Alexander',
    'email': 'albert.alexander@hussa.rs'
  },
  {
    'id': 2,
    'name': 'Philip',
    'lastname': 'Hughes',
    'email': 'philip.hughes@hussa.rs'
  },
  {
    'id': 3,
    'name': 'Walter',
    'lastname': 'Nelson',
    'email': 'walter.nelson@hussa.rs'
  },
  {
    'id': 4,
    'name': 'Ashley',
    'lastname': 'Hernandez',
    'email': 'ashley.hernandez@hussa.rs'
  },
];
