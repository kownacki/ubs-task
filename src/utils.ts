import { Output } from './types';
import { CATEGORIES_PATH } from './constants';

export const formDataToOutput = (formData: FormData) => {
  const rawOutput:Record<string, string> = {};

  formData.forEach((value, key) => rawOutput[key] = value as string);

  // rawOutput has only strings inside and different structure than desired output. Thus we need to transform some of them.
  const isPaidEvent = rawOutput.paid_event === 'paid';
  const output: Output = {
    title: rawOutput.title,
    description: rawOutput.description,
    category_id: Number(rawOutput.category_id),
    paid_event: isPaidEvent,
    ...(isPaidEvent ? { event_fee: Number(rawOutput.event_fee) } : {}),
    reward: Number(rawOutput.reward),
    date: `${rawOutput.date_day}T${rawOutput.date_time}`,
    duration: Number(rawOutput.duration) * 3600,
    coordinator: { id: rawOutput.coordinator_id, email: rawOutput.coordinator_email }
  }
  return output;
}

export const fetchData = async (path: string) => {
  const response = await fetch(path);
  return response.ok
      ? { data: await response.json(), isSuccess: true }
      : { isSuccess: false };
};
