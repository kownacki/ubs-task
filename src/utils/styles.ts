// Here are kept all common style configuration and utils used across all components.

export const sectionBoxShadow = '0 0 8px rgba(0, 0, 0, 0.2)';
