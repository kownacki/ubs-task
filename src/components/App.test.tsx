import React from 'react';
import { mount } from 'enzyme';
import { App } from './App';
import { fetchData } from '../utils';
import { categoriesStub, coordinatorsStub } from '../testutils';

jest.mock('../utils');

describe('App', () => {
  it('renders correctly when all data downloads successfully', () => {
    (fetchData as jest.Mock)
      .mockReturnValueOnce(Promise.resolve({ data: categoriesStub, isSuccess: true }))
      .mockReturnValueOnce(Promise.resolve({ data: coordinatorsStub, isSuccess: true }))
    const wrapper = mount(<App />);
    expect(wrapper).toMatchSnapshot();
  });
});
