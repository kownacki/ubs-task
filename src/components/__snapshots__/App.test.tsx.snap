// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly when all data downloads successfully 1`] = `
<App>
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <Header>
      <styled.div>
        <div
          className="sc-AxjAm ipFACa"
        >
          <styled.h1>
            <h1
              className="sc-AxirZ bsLAid"
            >
              New event
            </h1>
          </styled.h1>
        </div>
      </styled.div>
    </Header>
    <Contents>
      <styled.div>
        <div
          className="sc-fzoLag jKtBLq"
        >
          <styled.form
            id="form"
            onSubmit={[Function]}
          >
            <form
              className="sc-fzoXzr kSzmVC"
              id="form"
              onSubmit={[Function]}
            >
              <AboutSection>
                <Section
                  title="About"
                >
                  <styled.div>
                    <div
                      className="sc-AxiKw czhYLI"
                    >
                      <styled.div>
                        <div
                          className="sc-AxhCb wofSo"
                        >
                          <styled.div>
                            <div
                              className="sc-AxhUy iTOaem"
                            >
                              <styled.h2>
                                <h2
                                  className="sc-AxgMl dAMAgz"
                                >
                                  About
                                </h2>
                              </styled.h2>
                            </div>
                          </styled.div>
                          <styled.div>
                            <div
                              className="sc-AxheI hdbqHJ"
                            >
                              <FormLine
                                label="Title"
                                required={true}
                              >
                                <styled.div>
                                  <div
                                    className="sc-Axmtr cfmhAh"
                                  >
                                    <styled.div>
                                      <div
                                        className="sc-AxmLO hMNqXn"
                                      >
                                        Title *
                                      </div>
                                    </styled.div>
                                    <styled.div>
                                      <div
                                        className="sc-fzozJi eVevxM"
                                      >
                                        <styled.input
                                          name="title"
                                          placeholder="Make it short and clear"
                                          required={true}
                                          width="100%"
                                        >
                                          <input
                                            className="sc-fzpans bblOfU"
                                            name="title"
                                            placeholder="Make it short and clear"
                                            required={true}
                                            width="100%"
                                          />
                                        </styled.input>
                                      </div>
                                    </styled.div>
                                  </div>
                                </styled.div>
                              </FormLine>
                              <FormLine
                                label="Description"
                                required={true}
                              >
                                <styled.div>
                                  <div
                                    className="sc-Axmtr cfmhAh"
                                  >
                                    <styled.div>
                                      <div
                                        className="sc-AxmLO hMNqXn"
                                      >
                                        Description *
                                      </div>
                                    </styled.div>
                                    <styled.div>
                                      <div
                                        className="sc-fzozJi eVevxM"
                                      >
                                        <TextArea
                                          maxLength={140}
                                          name="description"
                                          placeholder="Write about your event, be creative"
                                          required={true}
                                          rows={8}
                                        >
                                          <div>
                                            <styled.textarea
                                              maxLength={140}
                                              name="description"
                                              onChange={[Function]}
                                              placeholder="Write about your event, be creative"
                                              required={true}
                                              rows={8}
                                            >
                                              <textarea
                                                className="sc-fzplWN cECDgO"
                                                maxLength={140}
                                                name="description"
                                                onChange={[Function]}
                                                placeholder="Write about your event, be creative"
                                                required={true}
                                                rows={8}
                                              />
                                            </styled.textarea>
                                            <styled.div>
                                              <div
                                                className="sc-fznyAO bAmZAn"
                                              >
                                                0 / 140
                                              </div>
                                            </styled.div>
                                          </div>
                                        </TextArea>
                                      </div>
                                    </styled.div>
                                  </div>
                                </styled.div>
                              </FormLine>
                              <Categories>
                                <FormLine
                                  label="Category"
                                >
                                  <styled.div>
                                    <div
                                      className="sc-Axmtr cfmhAh"
                                    >
                                      <styled.div>
                                        <div
                                          className="sc-AxmLO hMNqXn"
                                        >
                                          Category
                                        </div>
                                      </styled.div>
                                      <styled.div>
                                        <div
                                          className="sc-fzozJi eVevxM"
                                        >
                                          <Select
                                            name="category_id"
                                          >
                                            <styled.div>
                                              <div
                                                className="sc-fznKkj dGINIB"
                                              >
                                                <styled.select
                                                  name="category_id"
                                                >
                                                  <select
                                                    className="sc-fznZeY jWuLvU"
                                                    name="category_id"
                                                  >
                                                    <option
                                                      hidden={true}
                                                      value=""
                                                    >
                                                      Select category
                                                    </option>
                                                  </select>
                                                </styled.select>
                                              </div>
                                            </styled.div>
                                          </Select>
                                        </div>
                                      </styled.div>
                                    </div>
                                  </styled.div>
                                </FormLine>
                              </Categories>
                              <Payment>
                                <FormLine
                                  label="Payment"
                                >
                                  <styled.div>
                                    <div
                                      className="sc-Axmtr cfmhAh"
                                    >
                                      <styled.div>
                                        <div
                                          className="sc-AxmLO hMNqXn"
                                        >
                                          Payment
                                        </div>
                                      </styled.div>
                                      <styled.div>
                                        <div
                                          className="sc-fzozJi eVevxM"
                                        >
                                          <styled.span>
                                            <span
                                              className="sc-fzoLsD goxRTH"
                                            >
                                              <styled.input
                                                defaultChecked={true}
                                                id="free"
                                                name="paid_event"
                                                onChange={[Function]}
                                                value="free"
                                              >
                                                <input
                                                  className="sc-fzokOt csOIHo"
                                                  defaultChecked={true}
                                                  id="free"
                                                  name="paid_event"
                                                  onChange={[Function]}
                                                  type="radio"
                                                  value="free"
                                                />
                                              </styled.input>
                                              <label
                                                htmlFor="free"
                                              >
                                                Free event
                                              </label>
                                            </span>
                                          </styled.span>
                                          <styled.span>
                                            <span
                                              className="sc-fzoLsD goxRTH"
                                            >
                                              <styled.input
                                                id="paid"
                                                name="paid_event"
                                                onChange={[Function]}
                                                value="paid"
                                              >
                                                <input
                                                  className="sc-fzokOt csOIHo"
                                                  id="paid"
                                                  name="paid_event"
                                                  onChange={[Function]}
                                                  type="radio"
                                                  value="paid"
                                                />
                                              </styled.input>
                                              <label
                                                htmlFor="paid"
                                              >
                                                Paid event
                                              </label>
                                            </span>
                                          </styled.span>
                                        </div>
                                      </styled.div>
                                    </div>
                                  </styled.div>
                                </FormLine>
                              </Payment>
                              <FormLine
                                label="Reward"
                              >
                                <styled.div>
                                  <div
                                    className="sc-Axmtr cfmhAh"
                                  >
                                    <styled.div>
                                      <div
                                        className="sc-AxmLO hMNqXn"
                                      >
                                        Reward
                                      </div>
                                    </styled.div>
                                    <styled.div>
                                      <div
                                        className="sc-fzozJi eVevxM"
                                      >
                                        <styled.input
                                          min={0}
                                          name="reward"
                                          placeholder="Number"
                                          type="number"
                                          width="90px"
                                        >
                                          <input
                                            className="sc-fzpans yGyEV"
                                            min={0}
                                            name="reward"
                                            placeholder="Number"
                                            type="number"
                                            width="90px"
                                          />
                                        </styled.input>
                                        <styled.span>
                                          <span
                                            className="sc-fzoLsD goxRTH"
                                          >
                                            reward points for attendance
                                          </span>
                                        </styled.span>
                                      </div>
                                    </styled.div>
                                  </div>
                                </styled.div>
                              </FormLine>
                            </div>
                          </styled.div>
                        </div>
                      </styled.div>
                    </div>
                  </styled.div>
                </Section>
              </AboutSection>
              <CoordinatorSection>
                <Section
                  title="Coordinator"
                >
                  <styled.div>
                    <div
                      className="sc-AxiKw czhYLI"
                    >
                      <styled.div>
                        <div
                          className="sc-AxhCb wofSo"
                        >
                          <styled.div>
                            <div
                              className="sc-AxhUy iTOaem"
                            >
                              <styled.h2>
                                <h2
                                  className="sc-AxgMl dAMAgz"
                                >
                                  Coordinator
                                </h2>
                              </styled.h2>
                            </div>
                          </styled.div>
                          <styled.div>
                            <div
                              className="sc-AxheI hdbqHJ"
                            >
                              <FormLine
                                label="Responsible"
                                required={true}
                              >
                                <styled.div>
                                  <div
                                    className="sc-Axmtr cfmhAh"
                                  >
                                    <styled.div>
                                      <div
                                        className="sc-AxmLO hMNqXn"
                                      >
                                        Responsible *
                                      </div>
                                    </styled.div>
                                    <styled.div>
                                      <div
                                        className="sc-fzozJi eVevxM"
                                      />
                                    </styled.div>
                                  </div>
                                </styled.div>
                              </FormLine>
                              <FormLine
                                label="Email"
                              >
                                <styled.div>
                                  <div
                                    className="sc-Axmtr cfmhAh"
                                  >
                                    <styled.div>
                                      <div
                                        className="sc-AxmLO hMNqXn"
                                      >
                                        Email
                                      </div>
                                    </styled.div>
                                    <styled.div>
                                      <div
                                        className="sc-fzozJi eVevxM"
                                      >
                                        <styled.input
                                          data-test="email-input"
                                          name="coordinator_email"
                                          placeholder="Email"
                                          required={true}
                                          type="email"
                                          width="100%"
                                        >
                                          <input
                                            className="sc-fzpans bblOfU"
                                            data-test="email-input"
                                            name="coordinator_email"
                                            placeholder="Email"
                                            required={true}
                                            type="email"
                                            width="100%"
                                          />
                                        </styled.input>
                                      </div>
                                    </styled.div>
                                  </div>
                                </styled.div>
                              </FormLine>
                            </div>
                          </styled.div>
                        </div>
                      </styled.div>
                    </div>
                  </styled.div>
                </Section>
              </CoordinatorSection>
              <WhenSection>
                <Section
                  title="When"
                >
                  <styled.div>
                    <div
                      className="sc-AxiKw czhYLI"
                    >
                      <styled.div>
                        <div
                          className="sc-AxhCb wofSo"
                        >
                          <styled.div>
                            <div
                              className="sc-AxhUy iTOaem"
                            >
                              <styled.h2>
                                <h2
                                  className="sc-AxgMl dAMAgz"
                                >
                                  When
                                </h2>
                              </styled.h2>
                            </div>
                          </styled.div>
                          <styled.div>
                            <div
                              className="sc-AxheI hdbqHJ"
                            >
                              <FormLine
                                label="Starts on"
                                required={true}
                              >
                                <styled.div>
                                  <div
                                    className="sc-Axmtr cfmhAh"
                                  >
                                    <styled.div>
                                      <div
                                        className="sc-AxmLO hMNqXn"
                                      >
                                        Starts on *
                                      </div>
                                    </styled.div>
                                    <styled.div>
                                      <div
                                        className="sc-fzozJi eVevxM"
                                      >
                                        <styled.input
                                          name="date_day"
                                          required={true}
                                          type="date"
                                        >
                                          <input
                                            className="sc-fzpans kSLHGr"
                                            name="date_day"
                                            required={true}
                                            type="date"
                                          />
                                        </styled.input>
                                        <styled.span>
                                          <span
                                            className="sc-fzoLsD goxRTH"
                                          >
                                            at
                                          </span>
                                        </styled.span>
                                        <styled.input
                                          name="date_time"
                                          required={true}
                                          type="time"
                                        >
                                          <input
                                            className="sc-fzpans kSLHGr"
                                            name="date_time"
                                            required={true}
                                            type="time"
                                          />
                                        </styled.input>
                                      </div>
                                    </styled.div>
                                  </div>
                                </styled.div>
                              </FormLine>
                              <FormLine
                                label="Duration"
                              >
                                <styled.div>
                                  <div
                                    className="sc-Axmtr cfmhAh"
                                  >
                                    <styled.div>
                                      <div
                                        className="sc-AxmLO hMNqXn"
                                      >
                                        Duration
                                      </div>
                                    </styled.div>
                                    <styled.div>
                                      <div
                                        className="sc-fzozJi eVevxM"
                                      >
                                        <styled.input
                                          min={0}
                                          name="duration"
                                          placeholder="Number"
                                          type="number"
                                          width="90px"
                                        >
                                          <input
                                            className="sc-fzpans yGyEV"
                                            min={0}
                                            name="duration"
                                            placeholder="Number"
                                            type="number"
                                            width="90px"
                                          />
                                        </styled.input>
                                        <styled.span>
                                          <span
                                            className="sc-fzoLsD goxRTH"
                                          >
                                            hour
                                          </span>
                                        </styled.span>
                                      </div>
                                    </styled.div>
                                  </div>
                                </styled.div>
                              </FormLine>
                            </div>
                          </styled.div>
                        </div>
                      </styled.div>
                    </div>
                  </styled.div>
                </Section>
              </WhenSection>
              <styled.div>
                <div
                  className="sc-fzpjYC kGPIwg"
                >
                  <styled.input
                    value="Publish Event"
                  >
                    <input
                      className="sc-fzqBZW gTTtbM"
                      type="submit"
                      value="Publish Event"
                    />
                  </styled.input>
                </div>
              </styled.div>
            </form>
          </styled.form>
        </div>
      </styled.div>
    </Contents>
  </Provider>
</App>
`;
